packages/map/src/map/hooks.jsx

```diff
@@ -1,65 +1,101 @@
-import { useEffect, useState, useRef, useContext } from 'react';
-import ReactMapboxGl from 'react-mapbox-gl';
+import { useEffect, useState, useContext, useMemo } from 'react';
+import { Map, NavigationControl } from 'mapbox-gl';
+
+import { getMapOptions, getBoundingBox, mergeChildrenBounds } from './utils';
 import { getMedia, media } from '../utils';
 import MapContext from './context';
 
 
-export const useLocked = (locked, lockedMobile) => {
-  const [isLocked, setLocked] = useState(locked);
+export const useMapContext = () => useContext(MapContext);
+
+export const useMapEffect = (fn, deps) => {
+  const { map } = useMapContext();
+  if (deps) deps.push(map);
 
   useEffect(() => {
-    const isMobile = !getMedia(global, media.mediaM);
-    setLocked(isMobile ? lockedMobile : locked);
-  }, [locked, lockedMobile]);
+    if (!map) return;
+
+    const destroy = fn(map);
 
-  return isLocked;
+    let isRemoved = false;
+    const removeHandler = () => { isRemoved = true; };
+    map.on('remove', removeHandler);
+
+    return () => {
+      if (!isRemoved) {
```

Comment:
what are we solving here? multiple calls? race condition? seems like a complicated workaround

Reply:
race condition

you have to listen to the map getting removed, otherwise it will throw as far as I know

Reply:
can we refactor this in a cleaner way or leave a comment?

Reply:
Wouldn't know how to do it cleaner